### PR TITLE
Manifest trie

### LINKED ISSUE #1777: Update the manifest parsing code to accommodate the new trie format
Following https://github.com/web-platform-tests/wpt/pull/16537 , we need to adjust our manifest-related code to accommodate the new, fundamentally different format.

Fortunately, all features that rely on `/api/manifest` are off by default. And even for the users who opted in, the failure mode is expected to be graceful.

### PR DESCRIPTION
**Pull request by Hexcles**
The first commit fixes #1777 , while the second commit implements a new API for `wpt-metadata` (cc @KyleJu : let me know if `Contains` isn't enough for wpt-metadata).

### CODE REVIEW

File: shared/manifest.go [hunk 1]
```diff
@@ -4,7 +4,69 @@
...
+)
+
+// Manifest represents a JSON blob of all the WPT tests.
+type Manifest struct {
+	Items   map[string]rawManifestTrie `json:"items,omitempty"`
+	Version *int                       `json:"version,omitempty"`
+	URLBase string                     `json:"url_base,omitempty"`
+}
+
+type rawManifestTrie json.RawMessage
```
[the remark this answers is not in the document]
Comment (Hexcles) [reply shared/manifest.go]: Using `json.RawMessage` is a performance optimization to delay parsing, while wrapping it in our custom type is to add methods to it (the Golang way of inheritance, kind of). I'll document this together with why we need to delegate marshal/unmarshal.

File: shared/manifest.go [hunk 2]
```diff
@@ -4,7 +4,69 @@
...
+func (t rawManifestTrie) MarshalJSON() ([]byte, error) {
+	return json.RawMessage(t).MarshalJSON()
+}
+
+func (t *rawManifestTrie) UnmarshalJSON(b []byte) error {
+	return json.Unmarshal(b, (*json.RawMessage)(t))
+}
+
+func (t rawManifestTrie) FilterByPath(pathParts []string) (rawManifestTrie, error) {
+	if t == nil || len(pathParts) == 0 {
```
[the remark this answers is not in the document]
Comment (Hexcles) [reply shared/manifest.go]: `rawManifestTrie` is a `json.RawMessage` which is itself a byte slice, so yes. Golang type check will complain if can't be `nil`.

File: shared/manifest.go [hunk 3]
```diff
@@ -4,7 +4,69 @@
...
+		parts := strings.Split(p, "/")
+		// Split always returns at least one element.
+		// Remove the leading empty part.
+		if parts[0] == "" {
+			parts = parts[1:]
+		}
+		for testType, trie := range m.Items {
+			filtered, err := trie.FilterByPath(parts)
+			if err != nil {
+				return result, err
```
[the remark this answers is not in the document]
Comment (Hexcles) [reply shared/manifest.go]: I'm not aware of a best practice regarding this and I was simply copying the original behaviour, but your argument for returning `nil` makes a lot of sense here so I'll do that.

File: shared/manifest.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (stephenmcgruer) [reply shared/manifest.go]: (Argh, github outages... I deleted this comment after learning more about what is going on here. But apparently that delete didn't go through -_-.)

**Pull request merged**